File: server/src/Tables/DbVector2.cs
namespace masks.server.Tables;

[SpacetimeDB.Type]
public partial struct DbVector2
{
    public float X;
    public float Y;

    public DbVector2(float x, float y)
    {
        X = x;
        Y = y;
    }

    public float SqrMagnitude => X * X + Y * Y;
    public float Magnitude => MathF.Sqrt(SqrMagnitude);
    public DbVector2 Normalized
    {
        get
        {
            var mag = Magnitude;
            if (mag > float.Epsilon)
                return this / mag;
            return new DbVector2(0, 0);
        }
    }

    public static DbVector2 operator +(DbVector2 a, DbVector2 b) => new(a.X + b.X, a.Y + b.Y);
    public static DbVector2 operator -(DbVector2 a, DbVector2 b) => new(a.X - b.X, a.Y - b.Y);
    public static DbVector2 operator *(DbVector2 a, float b) => new(a.X * b, a.Y * b);
    public static DbVector2 operator /(DbVector2 a, float b) => new(a.X / b, a.Y / b);
}

File: server/src/tables/dbvector2.rs
use std::ops::{Add, Sub, Mul, Div};
use spacetimedb::SpacetimeType;

#[derive(SpacetimeType, Debug, Clone, Copy, PartialEq)]
pub struct DbVector2 {
    pub x: f32,
    pub y: f32,
}

impl DbVector2 {
    pub fn new(x: f32, y: f32) -> Self {
        Self { x, y }
    }

    pub fn sqr_magnitude(&self) -> f32 {
        self.x * self.x + self.y * self.y
    }

    pub fn magnitude(&self) -> f32 {
        self.sqr_magnitude().sqrt()
    }

    pub fn normalized(&self) -> Self {
        let mag = self.magnitude();
        if mag > f32::EPSILON {
            *self / mag
        } else {
            Self::new(0.0, 0.0)
        }
    }
}

impl Add for DbVector2 {
    type Output = Self;
    fn add(self, other: Self) -> Self {
        Self::new(self.x + other.x, self.y + other.y)
    }
}

impl Sub for DbVector2 {
    type Output = Self;
    fn sub(self, other: Self) -> Self {
        Self::new(self.x - other.x, self.y - other.y)
    }
}

impl Mul<f32> for DbVector2 {
    type Output = Self;
    fn mul(self, scalar: f32) -> Self {
        Self::new(self.x * scalar, self.y * scalar)
    }
}

impl Div<f32> for DbVector2 {
    type Output = Self;
    fn div(self, scalar: f32) -> Self {
        Self::new(self.x / scalar, self.y / scalar)
    }
}
